Add ismip6 forcing testgroup

hollyhan opened the pull request:
This PR adds a test group to COMPASS that processes the ISMIP6 forcing data for the Antarctica. Processing includes remapping the original data onto MALI mesh and renaming variables so that the output data can be used to force MALI. 
Currently, three test cases are added to the test group, `atmosphere`, `ocean_basal` and `ocean_thermal`.

pep8speaks commented:
Hello @hollyhan! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`compass/landice/tests/ismip6_forcing/__init__.py`](https://github.com/MPAS-Dev/compass/blob/046dd7b24929053781e04809058b897d9de43103/compass/landice/tests/ismip6_forcing/__init__.py):

> [Line 17:1](https://github.com/MPAS-Dev/compass/blob/046dd7b24929053781e04809058b897d9de43103/compass/landice/tests/ismip6_forcing/__init__.py#L17): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

* In the file [`compass/landice/tests/ismip6_forcing/atmosphere/__init__.py`](https://github.com/MPAS-Dev/compass/blob/046dd7b24929053781e04809058b897d9de43103/compass/landice/tests/ismip6_forcing/atmosphere/__init__.py):

> [Line 6:1](https://github.com/MPAS-Dev/compass/blob/046dd7b24929053781e04809058b897d9de43103/compass/landice/tests/ismip6_forcing/atmosphere/__init__.py#L6): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 45:1](https://github.com/MPAS-Dev/compass/blob/046dd7b24929053781e04809058b897d9de43103/compass/landice/tests/ismip6_forcing/atmosphere/__init__.py#L45): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

* In the file [`compass/landice/tests/ismip6_forcing/atmosphere/create_mapfile_smb.py`](https://github.com/MPAS-Dev/compass/blob/046dd7b24929053781e04809058b897d9de43103/compass/landice/tests/ismip6_forcing/atmosphere/create_mapfile_smb.py):

> [Line 132:80](https://github.com/MPAS-Dev/compass/blob/046dd7b24929053781e04809058b897d9de43103/compass/landice/tests/ismip6_forcing/atmosphere/create_mapfile_smb.py#L132): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)

* In the file [`compass/landice/tests/ismip6_forcing/atmosphere/process_smb.py`](https://github.com/MPAS-Dev/compass/blob/046dd7b24929053781e04809058b897d9de43103/compass/landice/tests/ismip6_forcing/atmosphere/process_smb.py):

> [Line 7:80](https://github.com/MPAS-Dev/compass/blob/046dd7b24929053781e04809058b897d9de43103/compass/landice/tests/ismip6_forcing/atmosphere/process_smb.py#L7): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (97 > 79 characters)
> [Line 43:39](https://github.com/MPAS-Dev/compass/blob/046dd7b24929053781e04809058b897d9de43103/compass/landice/tests/ismip6_forcing/atmosphere/process_smb.py#L43): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals

* In the file [`compass/landice/tests/ismip6_forcing/ocean/__init__.py`](https://github.com/MPAS-Dev/compass/blob/046dd7b24929053781e04809058b897d9de43103/compass/landice/tests/ismip6_forcing/ocean/__init__.py):

> [Line 44:59](https://github.com/MPAS-Dev/compass/blob/046dd7b24929053781e04809058b897d9de43103/compass/landice/tests/ismip6_forcing/ocean/__init__.py#L44): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

* In the file [`compass/landice/tests/ismip6_forcing/ocean/create_mapfile.py`](https://github.com/MPAS-Dev/compass/blob/046dd7b24929053781e04809058b897d9de43103/compass/landice/tests/ismip6_forcing/ocean/create_mapfile.py):

> [Line 78:5](https://github.com/MPAS-Dev/compass/blob/046dd7b24929053781e04809058b897d9de43103/compass/landice/tests/ismip6_forcing/ocean/create_mapfile.py#L78): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)

* In the file [`compass/landice/tests/ismip6_forcing/ocean/process_basal_melt.py`](https://github.com/MPAS-Dev/compass/blob/046dd7b24929053781e04809058b897d9de43103/compass/landice/tests/ismip6_forcing/ocean/process_basal_melt.py):

> [Line 4:80](https://github.com/MPAS-Dev/compass/blob/046dd7b24929053781e04809058b897d9de43103/compass/landice/tests/ismip6_forcing/ocean/process_basal_melt.py#L4): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)

* In the file [`compass/landice/tests/ismip6_forcing/ocean/process_thermal_forcing.py`](https://github.com/MPAS-Dev/compass/blob/046dd7b24929053781e04809058b897d9de43103/compass/landice/tests/ismip6_forcing/ocean/process_thermal_forcing.py):

> [Line 5:80](https://github.com/MPAS-Dev/compass/blob/046dd7b24929053781e04809058b897d9de43103/compass/landice/tests/ismip6_forcing/ocean/process_thermal_forcing.py#L5): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)

trhille commented on compass/landice/tests/ismip6_forcing/atmosphere/process_smb.py:
```diff
@@ -0,0 +1,162 @@
+        ds = ds.rename(ismip6_to_mali_dims)
+
+        ismip6_to_mali_vars = dict(
+            smb_anomaly="sfcMassBal")
```
Need to add smb_anomaly to baseline SMB.

xylar commented on compass/landice/tests/ismip6_forcing/atmosphere/process_smb.py:
```diff
@@ -0,0 +1,336 @@
+
+        ds = xr.open_dataset(output_file)
+        ds_base = xr.open_dataset(mali_mesh_file)
+        for i in range(len(ds["sfcMassBal"])):
+            ds["sfcMassBal"][i] = ds["sfcMassBal"][i] \
+                                  + ds_base["sfcMassBal"][0, :]
```
If I understand right, `i` is the time index here?  If so, let's chat in a bit about how to do this in a slightly different way that it more typical for xarray data sets.

xylar replied:
This would be at least the rough idea of what I would do:
```suggestion
        # get the first time index
        ref_smb = ds_base["sfcMassBal"].isel(Time=0)
        # broadcast so time is the same size as in ds
        ref_smb = ref_smb.broadcast_like(ds["sfcMassBal"])
        ds["sfcMassBal"] = ds["sfcMassBal"] + ref_smb
```

hollyhan commented:
Hi @xylar, @matthewhoffman, @xylar, 

This PR is now ready to be formally reviewed. I feel slightly guilty of requesting review from all three of you especially when everyone is so busy with other ongoing work stuff, so feel free to adjust/share the workload as you like. Thank you!!

hollyhan commented:
Oh yes, thanks for the reminder, @xylar.
You can find the config file on Cori in the following path: `/global/cscratch1/sd/hollyhan/compass`

matthewhoffman commented on compass/landice/tests/ismip6_forcing/ocean_thermal/process_thermal_forcing.py:
```diff
@@ -0,0 +1,319 @@
+                            target=os.path.join(base_path_ismip6, input_file[0]))
+        self.add_output_file(filename=output_file)
+
+    def run(self):
```
When I run this test case, it seems like nothing happens.  It runs just a short time, there are no output files generated, and in the log I just see:
```
Loading conda environment
Done.

Loading Spack environment...
```
I'm not sure how to debug this, but I'm open to suggestions.  (cc: @xylar )

matthewhoffman reviewed (changes requested):
@hollyhan , I'm excited about having this processing workflow reproducible via compass.  This is going to be definitely worth the investment.  I have a number of review comments to get things started.  I was only able to successfully run the ocean_basal case of the three, but I'm not sure how best to debug the problems with the other two.  It may require a little team discussion to figure out.  Once I manage to get them all running and can look at the output, I might have other feedback, but there is enough in this review to get you started with some revisions.

matthewhoffman commented:
@hollyhan , per our conversation, we'll need to figure out if a base .cfg file is required as well.

hollyhan commented on docs/users_guide/landice/test_groups/ismip6_forcing.rst:
```diff
@@ -0,0 +1,111 @@
+    # Available scenarios are the following: RCP26, RCP85, SSP126, SSP585
+    scenario = RCP26-repeat
+
+    # name of the mali mesh used to name mapping files (e,g. Antarctica_8to80km). User has to supply.
```
It is used to name mapping files created in `build_mapping_files` (e,g. 'map_ismip6_8km_to_{mali_mesh_name}_{method_remap}.nc').  I've added the note on the doc.

hollyhan commented on compass/landice/tests/ismip6_forcing/atmosphere/process_smb.py:
```diff
@@ -0,0 +1,340 @@
+        # get the first time index
+        ref_smb = ds_base["sfcMassBal"].isel(Time=0)
+        # broadcast so time is the same size as in ds
+        ref_smb = ref_smb.broadcast(ds["sfcMassBal"])
```
This error is a weird one because I ran into it once, but then it disappeared at another time, which is why I committed the changes. This was something that I think @xylar was also confused about and thought there was nothing wrong with the line of the code itself and thus shouldn't give an error. Good news is that this line is not really necessary and the code runs as expected without it. I've taken the line out, and the testcase should work all fine now.

hollyhan replied:
Turns out the syntax we were using were not quite correct. Still puzzling though why I thought it worked for me. I must have tested in a wrong branch that didn't have the line `ref_smb = ref_smb.broadcast(ds["sfcMassBal"])`. Anyhow, we still don't need the line as it is functionally redundant.

hollyhan commented:
Thanks for the quick and helpful review, @matthewhoffman! I've addressed all of your review comments and requested another review. Everything works fine at least for me. Hopefully the smb test case that failed for you works now. I'm not sure why you didn't get any outputs nor error for one of the ocean test cases. @xylar and @trhille, it would be really helpful to know if you also run into this issue and whether you get any errors when/if you do.  Thank you!

hollyhan commented on compass/landice/tests/ismip6_forcing/atmosphere/process_smb.py:
```diff
@@ -0,0 +1,340 @@
+
+        # place the output file in appropriate directory
+        if output_base_path == "NotAvailable":
+            return
```
Alternatively, I'll write out some messages using `logger. warn('')`

hollyhan commented on compass/landice/tests/ismip6_forcing/atmosphere/process_smb.py:
```diff
@@ -0,0 +1,340 @@
+        # check if a mapfile exists
+        mapping_file = f"map_ismip6_8km_to_{mali_mesh_name}_{method_remap}.nc"
+
+        if not os.path.exists(mapping_file):
```
I got rid of the check in the function `build_mapping_file`.

hollyhan commented on compass/landice/tests/ismip6_forcing/ocean_basal/create_mapfile.py:
```diff
@@ -0,0 +1,83 @@
+    subprocess.check_call(args)
+
+    # make sure the input file uses the longitude convention of [0 2pi]
+    args = ["set_lat_lon_fields_in_planar_grid.py",
```
Update: After discussing with @xylar, we settled on removing the part. I've checked that the `set_lat_lon_fields_in_planar_grid.py` is called in the AIS-mesh generation testcase, which makes the call here redundant and unnecessary.

matthewhoffman replied:
Thanks for looking into that.  The solution you came up with makes sense to me.

xylar commented on compass/landice/tests/ismip6_forcing/atmosphere/create_mapfile_smb.py:
```diff
@@ -0,0 +1,136 @@
+    print("Removing the temporary scripfiles...")
+    os.remove(ismip6_scripfile)
+    os.remove(mali_scripfile)
+
+def create_atm_scrip(ismip6_grid_file, ismip6_scripfile):
```
Two blank lines here to the PEP8 bot is happy...

xylar commented on compass/landice/tests/ismip6_forcing/__init__.py:
```diff
@@ -0,0 +1,22 @@
+
+class Ismip6Forcing(TestGroup):
+    """
+    A test group for processing the ISMIP6
+         ocean and atmosphere forcing data
```
```suggestion
    A test group for processing the ISMIP6
    ocean and atmosphere forcing data
```

xylar commented on compass/landice/tests/ismip6_forcing/__init__.py:
```diff
@@ -0,0 +1,22 @@
+
+    def __init__(self, mpas_core):
+        """
+        mpas_core : compass.landice.Landice
+            the MPAS core that this test group belongs to
```
```suggestion
        Create the test group

        Parameters
        ----------
        mpas_core : compass.landice.Landice
            the MPAS core that this test group belongs to
```

xylar commented on compass/landice/tests/ismip6_forcing/atmosphere/__init__.py:
```diff
@@ -0,0 +1,58 @@
+        """
+        Configures test case
+        """
+        base_path_ismip6 = self.config.get(section="ismip6_ais",
+                                           option="base_path_ismip6")
+        base_path_mali = self.config.get(section="ismip6_ais",
+                                         option="base_path_mali")
+        mali_mesh_name = self.config.get(section="ismip6_ais",
+                                         option="mali_mesh_name")
+        mali_mesh_file = self.config.get(section="ismip6_ais",
+                                         option="mali_mesh_file")
+
+        if base_path_ismip6 == "NotAvailable":
+            raise ValueError("You need to supply a user config file, which "
+                             "should contain the ismip6_ais "
+                             "section with the base_path_ismip6 option")
+        if base_path_mali == "NotAvailable":
+            raise ValueError("You need to supply a user config file, which "
+                             "should contain the ismip6_ais "
+                             "section with the base_path_mali option")
+        if mali_mesh_name == "NotAvailable":
+            raise ValueError("You need to supply a user config file, which "
+                             "should contain the ismip6_ais "
+                             "section with the mali_mesh_name option")
+        if mali_mesh_file == "NotAvailable":
+            raise ValueError("You need to supply a user config file, which "
+                             "should contain the ismip6_ais "
+                             "section with the mali_mesh_file option")
```
It would be a little better to do this efficiently in a loop:
```suggestion
        section = 'ismip6_ais'
        for option in ['base_path_ismip6', 'base_path_mali', 'mali_mesh_name',
                       'mali_mesh_file']:
            value = self.config.get(section=section, option=option)
            if value == "NotAvailable":
                raise ValueError(f"You need to supply a user config file, which "
                                 f"should contain the {section} "
                                 f"section with the {option} option")
```

xylar replied:
Should this also check `period_endyear`, `model` and `scenario`?

xylar commented on compass/landice/tests/ismip6_forcing/atmosphere/create_mapfile_smb.py:
```diff
@@ -0,0 +1,136 @@
+from mpas_tools.scrip.from_mpas import scrip_from_mpas
+
+
+# function that creates a mapping file from ismip6 grid to mali mesh
+def build_mapping_file(ismip6_grid_file, mapping_file, mali_mesh_file=None,
+                       method_remap=None):
```
Comments like this one should instead be put in the docstring.

xylar commented on compass/landice/tests/ismip6_forcing/create_mapfile.py:
```diff
@@ -0,0 +1,76 @@
+from mpas_tools.scrip.from_mpas import scrip_from_mpas
+
+
+# function that creates a mapping file from ismip6 grid to mali mesh
```
Again, probably move this to the docstring.

xylar commented on compass/landice/tests/ismip6_forcing/ocean_basal/__init__.py:
```diff
@@ -0,0 +1,57 @@
+        step = ProcessBasalMelt(test_case=self)
+        self.add_step(step)
+
+    def configure(self):
+        """
+        Configures test case
+        """
+        base_path_ismip6 = self.config.get(section="ismip6_ais",
+                                           option="base_path_ismip6")
+        base_path_mali = self.config.get(section="ismip6_ais",
+                                         option="base_path_mali")
+        mali_mesh_name = self.config.get(section="ismip6_ais",
+                                         option="mali_mesh_name")
+        mali_mesh_file = self.config.get(section="ismip6_ais",
+                                         option="mali_mesh_file")
+
+        if base_path_ismip6 == "NotAvailable":
+            raise ValueError("You need to supply a user config file, which "
+                             "should contain the ismip6_ais "
+                             "section with the base_path_ismip6 option")
+        if base_path_mali == "NotAvailable":
+            raise ValueError("You need to supply a user config file, which "
+                             "should contain the ismip6_ais "
+                             "section with the base_path_mali option")
+        if mali_mesh_name == "NotAvailable":
+            raise ValueError("You need to supply a user config file, which "
+                             "should contain the ismip6_ais "
+                             "section with the mali_mesh_name option")
+        if mali_mesh_file == "NotAvailable":
+            raise ValueError("You need to supply a user config file, which "
+                             "should contain the ismip6_ais "
+                             "section with the mali_mesh_file option")
```
Could be a loop, like my suggestion above.  Since both test cases check the same config options, this could also be a shared function at the test-group level.

xylar commented on compass/landice/tests/ismip6_forcing/ocean_basal/process_basal_melt.py:
```diff
@@ -0,0 +1,230 @@
+        ds["ismip6shelfMelt_basin"] = ds_basin.basinNumber
+        write_netcdf(ds, combined_file_temp)
+
+    def remap_ismip6BasalMelt_to_mali(self, input_file, output_file,
```
I'd really strongly prefer we avoid mixed case in method names.
 ```suggestion
     def remap_ismip6_basal_melt_to_mali(self, input_file, output_file,
```

xylar commented on compass/landice/tests/ismip6_forcing/ocean_basal/process_basal_melt.py:
```diff
@@ -0,0 +1,230 @@
+
+        subprocess.check_call(args)
+
+    def rename_ismip6BasalMelt_to_mali_vars(self, remapped_file_temp,
```
```suggestion
    def rename_ismip6_basal_melt_to_mali_vars(self, remapped_file_temp,
```

xylar commented on compass/landice/tests/ismip6_forcing/ocean_thermal/__init__.py:
```diff
@@ -0,0 +1,57 @@
+        """
+        Configures test case
+        """
+        base_path_ismip6 = self.config.get(section="ismip6_ais",
+                                           option="base_path_ismip6")
+        base_path_mali = self.config.get(section="ismip6_ais",
+                                         option="base_path_mali")
+        mali_mesh_name = self.config.get(section="ismip6_ais",
+                                         option="mali_mesh_name")
+        mali_mesh_file = self.config.get(section="ismip6_ais",
+                                         option="mali_mesh_file")
+
+        if base_path_ismip6 == "NotAvailable":
+            raise ValueError("You need to supply a user config file, which "
+                             "should contain the ismip6_ais "
+                             "section with the base_path_ismip6 option")
+        if base_path_mali == "NotAvailable":
+            raise ValueError("You need to supply a user config file, which "
+                             "should contain the ismip6_ais "
+                             "section with the base_path_mali option")
+        if mali_mesh_name == "NotAvailable":
+            raise ValueError("You need to supply a user config file, which "
+                             "should contain the ismip6_ais "
+                             "section with the mali_mesh_name option")
+        if mali_mesh_file == "NotAvailable":
+            raise ValueError("You need to supply a user config file, which "
+                             "should contain the ismip6_ais "
+                             "section with the mali_mesh_file option")
```
As in previous 2 cases, I suggest calling a shared function and using a loop to make this more efficient.

xylar commented on compass/landice/tests/ismip6_forcing/ocean_thermal/process_thermal_forcing.py:
```diff
@@ -0,0 +1,319 @@
+        section = config['ismip6_ais_ocean_thermal']
+        process_obs_data = section.getboolean("process_obs_data")
+
+        if period_endyear == "NotAvailable":
+            raise ValueError("You need to supply a user config file, which "
+                             "should contain the ismip6_ais "
+                             "section with the period_endyear option")
+        if model == "NotAvailable":
+            raise ValueError("You need to supply a user config file, which "
+                             "should contain the ismip6_ais "
+                             "section with the model option")
+        if scenario == "NotAvailable":
+            raise ValueError("You need to supply a user config file, which "
+                             "should contain the ismip6_ais "
+                             "section with the scenario option")
```
This could be checked in the test case's `configure()` along with the other options (in a shared function).

xylar commented on compass/landice/tests/ismip6_forcing/ocean_thermal/process_thermal_forcing.py:
```diff
@@ -0,0 +1,319 @@
+        dst = os.path.join(output_path, output_file)
+        shutil.copy(src, dst)
+
+    def remap_ismip6thermalforcing_to_mali(self, input_file,
```
This is all lowercase, which I like but I would use underscores to separate words:
```suggestion
    def remap_ismip6_thermal_forcing_to_mali(self, input_file,
```

xylar commented on compass/landice/tests/ismip6_forcing/ocean_thermal/process_thermal_forcing.py:
```diff
@@ -0,0 +1,319 @@
+
+        subprocess.check_call(args)
+
+    def rename_ismip6thermalforcing_to_mali_vars(self, remapped_file_temp,
```
```suggestion
    def rename_ismip6_thermal_forcing_to_mali_vars(self, remapped_file_temp,
```

xylar commented on docs/developers_guide/landice/api.rst:
```diff
@@ -231,6 +231,25 @@ hydro_radial
+.. autosummary::
+   :toctrr: generate/
+   
+   Ismip6Forcing
+
+   atmosphere.Atmosphere
+   atmosphere.Atmosphere.run
+
+   ocean_basal.OceanBasal
+   ocean_basal.OceanBasal.run
+   
+   ocean_thermal.OceanThermal
+   ocean_thermal.OceanThermal.run
```
I think there are several more functions and methods that need to be added to the API.  Let me know if you need help with that.

xylar commented on docs/users_guide/landice/test_groups/ismip6_forcing.rst:
```diff
@@ -0,0 +1,109 @@
+config options
+--------------
+
+All three test cases share some set of default config options under the section
+``[ismip6_ais]`` and separate config options under the corresponding section:
```
What is shown here are not the default config options, they are examples of modifying the config options like one might do in a user config file.  This example is important but it would be better to first show the default config options and then show a typical user config file.

xylar commented on docs/users_guide/landice/test_groups/ismip6_forcing.rst:
```diff
@@ -0,0 +1,109 @@
+ocean_basal
+------------
+
+The ``landice.tests.ismip6_forcing.ocean_basal``
```
```suggestion
The ``landice/ismip6_forcing/ocean_basal``
```

xylar reviewed (changes requested):
@hollyhan, I'm going to try out this test case, but did a first pass through the code just looking for things to clean up.  You did a really superb job on this test group!  There are just a few things that could make it a little cleaner and the documentation a little easier for a user to figure out.

xylar commented on docs/users_guide/landice/test_groups/ismip6_forcing.rst:
```diff
@@ -0,0 +1,109 @@
+ismip6_forcing
+============
+
+The ``landice/ismip6_forcing`` test group implements processing of atmospheric
+and ocean forcing data of the Ice Sheet Model Intercomparison for CMIP6
+(ISMIP6) protocol. The ISMIP6 source data can be obtained by directly
+contacting the authors. Reference: https://www.climate-cryosphere.org/wiki/index.php?title=ISMIP6-Projections2300-Antarctica#A2.2_Retrieving_datasets_and_uploading_your_model_output
```
I think you should add something here about how a user needs to download the ISMIP6 forcing data in order to use these test cases.  As I understand it, they need to create a base directory that is equivalent to one of the GHub globus endpoints (`GHub-ISMIP6-Forcing` or `ISMIP6 Projections Forcing 2300`).  Then, they need to manually download the files in the same directory structure as on the endpoint.  How do they know what files to download for a given model, scenario, and period end year?  Could you give an example for 2100 and another for 2300?

xylar replied:
`output_base_path` also needs to be discussed.

xylar replied:
Don't forget to mention the files in `parameterizations` and `imbie2_basin_numbers_8km.nc`

xylar commented:
@hollyhan, the example config file in the documentation seems to be for your laptop.  Could you paste the contents of the config file you used on Cori here so I can use it?  Specifically, I'm trying to figure out where to get the MALI mesh file.

Also could you add that as well to the documentation?  How does a user get an appropriate MALI mesh file?

xylar commented on compass/landice/tests/ismip6_forcing/ismip6.cfg:
```diff
@@ -0,0 +1,51 @@
+# config options for ismip6 antarctic ice sheet data set
+[ismip6_ais]
```
This file needs to be called `ismip6_forcing.cfg` to be included automatically.  If you want to keep the current name, you will need to add it explicitly to each test case in its `configure()` method.

xylar replied:
When I don't supply a user config file, I'm seeing:
```
Traceback (most recent call last):
  File "/home/ac.xylar/anvil/miniconda3/envs/compass_ismip6_forcing/bin/compass", line 33, in <module>
    sys.exit(load_entry_point('compass', 'console_scripts', 'compass')())
  File "/gpfs/fs1/home/ac.xylar/mpas-work/compass/add_ismip6_forcing_testgroup/compass/__main__.py", line 63, in main
    commands[args.command]()
  File "/gpfs/fs1/home/ac.xylar/mpas-work/compass/add_ismip6_forcing_testgroup/compass/setup.py", line 372, in main
    setup_cases(tests=tests, numbers=args.case_num,
  File "/gpfs/fs1/home/ac.xylar/mpas-work/compass/add_ismip6_forcing_testgroup/compass/setup.py", line 146, in setup_cases
    setup_case(path, test_case, config_file, machine, work_dir,
  File "/gpfs/fs1/home/ac.xylar/mpas-work/compass/add_ismip6_forcing_testgroup/compass/setup.py", line 244, in setup_case
    test_case.configure()
  File "/gpfs/fs1/home/ac.xylar/mpas-work/compass/add_ismip6_forcing_testgroup/compass/landice/tests/ismip6_forcing/atmosphere/__init__.py", line 34, in configure
    base_path_ismip6 = self.config.get(section="ismip6_ais",
  File "/home/ac.xylar/anvil/miniconda3/envs/compass_ismip6_forcing/lib/python3.9/site-packages/mpas_tools/config.py", line 111, in get
    return self.combined.get(section, option)
  File "/home/ac.xylar/anvil/miniconda3/envs/compass_ismip6_forcing/lib/python3.9/configparser.py", line 781, in get
    d = self._unify_values(section, vars)
  File "/home/ac.xylar/anvil/miniconda3/envs/compass_ismip6_forcing/lib/python3.9/configparser.py", line 1152, in _unify_values
    raise NoSectionError(section) from None
configparser.NoSectionError: No section: 'ismip6_ais'
```
which clued me in to this being missing.

xylar replied:
I don't see `output_base_path` here.  The default behavior seems to be to write things to a local path called `None`.

xylar commented on compass/landice/tests/ismip6_forcing/atmosphere/create_mapfile_smb.py:
```diff
@@ -0,0 +1,136 @@
+                         "--method. Available options are 'bilinear',"
+                         "'neareststod', 'conserve'.")
+
+    args = ["ESMF_RegridWeightGen",
+            "-s", ismip6_scripfile,
+            "-d", mali_scripfile,
+            "-w", mapping_file,
+            "-m", method_remap,
+            "-i", "-64bit_offset",
+            "--dst_regional", "--src_regional"]
```
On Anvil, I'm getting errors when running this because it needs to be called with `srun`:
```
Abort(1091087) on node 0 (rank 0 in comm 0): Fatal error in PMPI_Init: Other MPI error, error stack:
MPIR_Init_thread(136): 
MPID_Init(950).......: 
MPIR_pmi_init(163)...: PMI_Init returned 14
Traceback (most recent call last):
  File "/home/ac.xylar/anvil/miniconda3/envs/compass_ismip6_forcing/bin/compass", line 33, in <module>
    sys.exit(load_entry_point('compass', 'console_scripts', 'compass')())
  File "/gpfs/fs1/home/ac.xylar/mpas-work/compass/add_ismip6_forcing_testgroup/compass/__main__.py", line 63, in main
    commands[args.command]()
  File "/gpfs/fs1/home/ac.xylar/mpas-work/compass/add_ismip6_forcing_testgroup/compass/run/serial.py", line 301, in main
    run_step(args.step_is_subprocess)
  File "/gpfs/fs1/home/ac.xylar/mpas-work/compass/add_ismip6_forcing_testgroup/compass/run/serial.py", line 262, in run_step
    test_case.run()
  File "/gpfs/fs1/home/ac.xylar/mpas-work/compass/add_ismip6_forcing_testgroup/compass/testcase.py", line 166, in run
    self._run_step(step, self.new_step_log_file)
  File "/gpfs/fs1/home/ac.xylar/mpas-work/compass/add_ismip6_forcing_testgroup/compass/testcase.py", line 282, in _run_step
    step.run()
  File "/gpfs/fs1/home/ac.xylar/mpas-work/compass/add_ismip6_forcing_testgroup/compass/landice/tests/ismip6_forcing/atmosphere/process_smb.py", line 106, in run
    self.remap_ismip6smb_to_mali(combined_file_temp, remapped_file_temp,
  File "/gpfs/fs1/home/ac.xylar/mpas-work/compass/add_ismip6_forcing_testgroup/compass/landice/tests/ismip6_forcing/atmosphere/process_smb.py", line 167, in remap_ismip6smb_to_mali
    build_mapping_file(input_file, mapping_file, mali_mesh_file,
  File "/gpfs/fs1/home/ac.xylar/mpas-work/compass/add_ismip6_forcing_testgroup/compass/landice/tests/ismip6_forcing/atmosphere/create_mapfile_smb.py", line 63, in build_mapping_file
    subprocess.check_call(args)
  File "/home/ac.xylar/anvil/miniconda3/envs/compass_ismip6_forcing/lib/python3.9/subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['ESMF_RegridWeightGen', '-s', 'temp_ismip6_8km_scrip.nc', '-d', 'temp_mali_scrip.nc', '-w', 'map_ismip6_8km_to_Antarctica_8to80km_bilinear.nc', '-m', 'bilinear', '-i', '-64bit_offset', '--dst_regional', '--src_regional']' returned non-zero exit status 1.
(compass_ismip6_forcing) [ac.xylar@b665 process_smb]$ which ESMF_RegridWeightGen
/lcrc/soft/climate/compass/anvil/spack/spack_for_mache_1.4.1/var/spack/environments/dev_compass_1_1_0-alpha_4_intel_impi/.spack-env/view/bin/ESMF_RegridWeightGen
```
Let's talk through how to do this.  It would also be an opportunity to set `cores` and `min_cores` and use a parallel call to `ESMF_RegridWeightGen`.

xylar replied:
The same will apply in `create_mapfile.py`.

xylar commented on compass/landice/tests/ismip6_forcing/ocean_basal/process_basal_melt.py:
```diff
@@ -0,0 +1,230 @@
+            if output_base_path == "NotAvailable":
+                return
+
+            output_path = f'{output_base_path}/basal_melt/parametrizations/'
```
Either `parameterizations` or `parametrizations` would be correct but the directory on GHub is `parameterizations` (with an extra `e`).

```suggestion
            output_path = f'{output_base_path}/basal_melt/parameterizations/'
```

xylar commented on compass/landice/tests/ismip6_forcing/ocean_basal/process_basal_melt.py:
```diff
@@ -0,0 +1,230 @@
+    # input files: input uniform melt rate coefficient (gamma0)
+    # and temperature correction per basin
+    _file_basin = "AIS/Ocean_Forcing/imbie2/imbie2_basin_numbers_8km.nc"
+    _files_coeff = {"AIS/Ocean_Forcing/parametrizations/coeff_gamma0_DeltaT_quadratic_local_5th_pct_PIGL_gamma_calibration.nc",
+                    "AIS/Ocean_Forcing/parametrizations/coeff_gamma0_DeltaT_quadratic_local_5th_percentile.nc",
+                    "AIS/Ocean_Forcing/parametrizations/coeff_gamma0_DeltaT_quadratic_local_95th_pct_PIGL_gamma_calibration.nc",
+                    "AIS/Ocean_Forcing/parametrizations/coeff_gamma0_DeltaT_quadratic_local_95th_percentile.nc",
+                    "AIS/Ocean_Forcing/parametrizations/coeff_gamma0_DeltaT_quadratic_local_median_PIGL_gamma_calibration.nc",
+                    "AIS/Ocean_Forcing/parametrizations/coeff_gamma0_DeltaT_quadratic_local_median.nc",
+                    "AIS/Ocean_Forcing/parametrizations/coeff_gamma0_DeltaT_quadratic_non_local_5th_pct_PIGL_gamma_calibration.nc",
+                    "AIS/Ocean_Forcing/parametrizations/coeff_gamma0_DeltaT_quadratic_non_local_5th_percentile.nc",
+                    "AIS/Ocean_Forcing/parametrizations/coeff_gamma0_DeltaT_quadratic_non_local_95th_pct_PIGL_gamma_calibration.nc",
+                    "AIS/Ocean_Forcing/parametrizations/coeff_gamma0_DeltaT_quadratic_non_local_95th_percentile.nc",
+                    "AIS/Ocean_Forcing/parametrizations/coeff_gamma0_DeltaT_quadratic_non_local_median_PIGL_gamma_calibration.nc",
+                    "AIS/Ocean_Forcing/parametrizations/coeff_gamma0_DeltaT_quadratic_non_local_median.nc"}
```
These all need to be in `parameterizations`, not `parametrizations`.

xylar reviewed (changes requested):
A few more things came up as I was testing out the test group.

With a few fixes, things will be working great!  Amazing work, @hollyhan!  Other than the tedium of downloading the individual forcing files, this is going to make it so easy to process all the forcing data we need.

For reference, here was my user config file on Anvil:
``` cfg
# config options for ismip6 antarctic ice sheet data set
[ismip6_ais]

# Base path to the input ismip6 ocean and smb forcing files. User has to supply.
base_path_ismip6 = /lcrc/group/e3sm/ac.xylar/ismip6_forcing/2100/

# Base path to the the MALI mesh. User has to supply.
base_path_mali = /lcrc/group/e3sm/ac.xylar/MALI_meshes

output_base_path = /lcrc/group/e3sm/ac.xylar/ismip6_forcing/MALI/

# Forcing end year of the ISMIP6 data. User has to supply.
# Available end years are 2100 and 2300.
period_endyear = 2100

# Name of climate model name used to generate ISMIP6 forcing data. User has to supply.
# Available model names are the following: CCSM4, CESM2-WACCM, CSIRO-Mk3-6-0, HadGEM2-ES, NorESM1-M, UKESM1-0-LL
model = UKESM1-0-LL

# Scenarios used by climate model. User has to supply.
# Available scenarios are the following: RCP26, RCP26-repeat, RCP85, SSP126, SSP585
scenario = SSP585

# name of the mali mesh. User has to supply. Note: It is used to name mapping files
# (e,g. 'map_ismip6_8km_to_{mali_mesh_name}_{method_remap}.nc').
mali_mesh_name = Antarctica_8to80km

# MALI mesh file to be used to build mapping file (netCDF format). User has to supply.
mali_mesh_file = Antarctica_8to80km_20220407.nc


# config options for ismip6 antarctic ice sheet SMB forcing data test cases
[ismip6_ais_atmosphere]

# Remapping method used in building a mapping file. Options include: bilinear, neareststod, conserve
method_remap = bilinear

# config options for ismip6 ocean thermal forcing data test cases
[ismip6_ais_ocean_thermal]

# Remapping method used in building a mapping file. Options include: bilinear, neareststod, conserve
method_remap = bilinear

# Set to True if the want to process observational thermal forcing data. Set to False if want to process model thermal forcing data.
# Note: when set True, the ['ismip6_ais'] config options 'period_endyear', 'model' and 'scenario' will be ignored.
process_obs_data = False


# config options for ismip6 ocean basal test case
[ismip6_ais_ocean_basal]

# Remapping method used in building a mapping file. Ocean basal testcase will always want to
# use neareststod method.
method_remap = neareststod


```

xylar commented on compass/landice/tests/ismip6_forcing/atmosphere/create_mapfile_smb.py:
```diff
@@ -0,0 +1,165 @@
+            "--file", mali_mesh_copy,
+            "--proj", ismip6_projection]
+
+    subprocess.check_call(args)
```
Here and for other calls like it, this needs to use `mpas_tools.logging.check_call` and pass `self.logger`.  Otherwise, output goes to the terminal instead of the log file.

xylar commented on compass/landice/tests/ismip6_forcing/atmosphere/process_smb.py:
```diff
@@ -0,0 +1,331 @@
+                "-m", mapping_file,
+                "-v", "smb_anomaly"]
+
+        subprocess.check_call(args)
```
Again, this needs to use `mpas_tools.logging.check_call` and pass `self.logger`.  Otherwise, output goes to the terminal instead of the log file.

xylar commented on compass/landice/tests/ismip6_forcing/create_mapfile.py:
```diff
@@ -0,0 +1,102 @@
+            "--proj", ismip6_projection,
+            "--rank", "2"]
+
+    subprocess.check_call(args)
```
Again, this needs to use `mpas_tools.logging.check_call` and pass `self.logger`.  Otherwise, output goes to the terminal instead of the log file.

xylar commented on compass/landice/tests/ismip6_forcing/create_mapfile.py:
```diff
@@ -0,0 +1,102 @@
+            "--file", mali_mesh_copy,
+            "--proj", ismip6_projection]
+
+    subprocess.check_call(args)
```
Again, this needs to use `mpas_tools.logging.check_call` and pass `self.logger`.  Otherwise, output goes to the terminal instead of the log file.

xylar commented on compass/landice/tests/ismip6_forcing/ocean_basal/process_basal_melt.py:
```diff
@@ -0,0 +1,232 @@
+                "-o", output_file,
+                "-m", mapping_file]
+
+        subprocess.check_call(args)
```
Again, this needs to use `mpas_tools.logging.check_call` and pass `self.logger`.  Otherwise, output goes to the terminal instead of the log file.

xylar commented on compass/landice/tests/ismip6_forcing/ocean_thermal/process_thermal_forcing.py:
```diff
@@ -0,0 +1,308 @@
+                "-o", output_file,
+                "-m", mapping_file]
+
+        subprocess.check_call(args)
```
Again, this needs to use `mpas_tools.logging.check_call` and pass `self.logger`.  Otherwise, output goes to the terminal instead of the log file.

xylar commented on compass/landice/tests/ismip6_forcing/ocean_thermal/process_thermal_forcing.py:
```diff
@@ -0,0 +1,308 @@
+            The test case this step belongs to
+        """
+        super().__init__(test_case=test_case, name='process_thermal_forcing',
+                         cores = 4, min_cores = 1)
```
```suggestion
                         cores=4, min_cores=1)
```
For keyword arguments like these, no spaces around the equals sign.

xylar commented on compass/landice/tests/ismip6_forcing/ocean_thermal/process_thermal_forcing.py:
```diff
@@ -0,0 +1,308 @@
+        """
+        Run this step of the test case
+        """
+        logger = self.logger
```
```suggestion
```
Currently, you don't use the logger here in the `run()` method.

xylar commented on compass/landice/tests/ismip6_forcing/create_mapfile.py:
```diff
@@ -0,0 +1,102 @@
+    print("Removing the temporary mesh and scripfiles...")
+    os.remove(ismip6_scripfile)
+    os.remove(mali_scripfile)
+    os.remove(mali_mesh_copy)
```
You need a blank line at the end of the file.

xylar reviewed (changes requested):
I'm noting a few changes here that need to be made and a few more on #410.  These should all be combined and fixed on this branch as I hope we can discuss today.

hollyhan commented:
Closing in favor of #410.
